services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/controller/GSController.java

```diff
@@ -0,0 +1,173 @@
+   * @return RestOperationResult in String
+   * @throws Exception Various exceptions that may occur during the create process
+   * @link https://github.com/haboy52581/alcor/blob/master/docs/modules/ROOT
+   *     /pages/infra_services/data_plane_manager.adoc
+   */
+  @PostMapping({"/subnet/", "v4/subnet/"})
+  @ResponseStatus(HttpStatus.CREATED)
+  public List<InternalDPMResult> createSubnet(@RequestBody NetworkConfiguration gs)
+      throws Exception {
+    gs.setOpType(Common.OperationType.CREATE);
+    gs.setRsType(Common.ResourceType.SUBNET);
+    return service(gs);
+  }
+
+  /**
+   * Accept north bound calls then transfer to ACA calls in southbound
+   *
+   * @param gs Encapsulation of NetworkConfiguration message
+   * @return RestOperationResult in String
+   * @throws Exception Various exceptions that may occur during the update process
+   * @link https://github.com/haboy52581/alcor/blob/master/docs/modules/ROOT
+   *     /pages/infra_services/data_plane_manager.adoc
+   */
+  @PutMapping({"/subnet/", "v4/subnet/"})
+  public List<InternalDPMResult> updateSubnet(@RequestBody NetworkConfiguration gs)
+      throws Exception {
+    gs.setOpType(Common.OperationType.UPDATE);
+    gs.setRsType(Common.ResourceType.SUBNET);
+    return service(gs);
+  }
+
+  /**
+   * Accept north bound calls then transfer to ACA calls in southbound
+   *
+   * @param gs Encapsulation of NetworkConfiguration message
+   * @return RestOperationResult in String
+   * @throws Exception Various exceptions that may occur during the delete process
+   * @link https://github.com/haboy52581/alcor/blob/master/docs/modules/ROOT
+   *     /pages/infra_services/data_plane_manager.adoc
+   */
+  @DeleteMapping({"/subnet/", "v4/subnet/"})
+  public List<InternalDPMResult> deleteSubnet(@RequestBody NetworkConfiguration gs)
+      throws Exception {
+    gs.setOpType(Common.OperationType.DELETE);
+    gs.setRsType(Common.ResourceType.SUBNET);
+    return service(gs);
+  }
+
+  private List<InternalDPMResult> service(NetworkConfiguration gs) {
+    // TODO: Create a verification framework for all resources
+    // leave isFast as true since SB GSinfo does not have fastpath attr
+    try {
+      return goalStateManager
+          .talkToACA(
+              goalStateManager.transformNorthToSouth(gs),
+              true,
+              Integer.parseInt(config.getPort()),
+              Boolean.valueOf(config.getOvs()))
+          .stream()
+          .flatMap(Collection::stream)
+          .map(
+              f -> {
+                return new InternalDPMResult(
+                    f.getResourceId(),
+                    f.getResourceType().toString(),
+                    f.getOperationStatus().toString(),
+                    f.getStateElapseTime());
+              })
+          .collect(Collectors.toList());
+    } catch (Exception e) {
+      e.printStackTrace();
+      InternalDPMResult t =
+          new InternalDPMResult(
```

Comment:
@haboy52581 Return a "handmade" results like this basically return 200 SUCCUSS to upstream microservice which is not expected.

Instead, we should look deep into possible exceptions, and use different error codes for different exceptions. 

One example is Port Manager which has a full list of home-grown exceptions https://github.com/futurewei-cloud/alcor/tree/master/services/port_manager/src/main/java/com/futurewei/alcor/portmanager/exception

Please check its implementation carefully and see if we can do the same here.

Reply:
fixed in 3f9c